File: be/app/schemas/badge.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel


class BadgeProgress(BaseModel):
	current: int
	target: int


class BadgeItem(BaseModel):
	id: uuid.UUID
	name: str
	description: str
	icon_url: str
	rarity: str  # common | rare | epic | legendary
	category: str
	criteria: dict[str, Any]
	is_hidden: bool
	is_unlocked: bool
	unlocked_at: datetime | None = None
	progress: BadgeProgress

	model_config = {"from_attributes": True}


class BadgeListResponse(BaseModel):
	data: list[BadgeItem]
	total: int


class FeaturedBadge(BaseModel):
	id: uuid.UUID
	name: str
	icon_url: str
	rarity: str
	unlocked_at: datetime | None = None

	model_config = {"from_attributes": True}


class FeaturedBadgeResponse(BaseModel):
	data: list[FeaturedBadge]

File: mobile/types/badge.ts
export type BadgeRarity = 'common' | 'rare' | 'epic' | 'legendary';

export type BadgeCategory =
  | 'quests'
  | 'social'
  | 'streak'
  | 'progression'
  | 'trust'
  | 'hidden';

export interface BadgeProgress {
  current: number;
  target: number;
}

export interface BadgeCriteria {
  type: string;
  target?: number;
  count?: number;
}

export interface BadgeItem {
  id: string;
  name: string;
  description: string;
  icon_url: string;
  rarity: BadgeRarity;
  category: string;
  criteria: BadgeCriteria;
  is_hidden: boolean;
  is_unlocked: boolean;
  unlocked_at: string | null;
  progress: BadgeProgress;
}

export interface FeaturedBadge {
  id: string;
  name: string;
  icon_url: string;
  rarity: BadgeRarity;
  unlocked_at: string | null;
}

export interface BadgeListResponse {
  data: BadgeItem[];
  total: number;
}

export interface FeaturedBadgeResponse {
  data: FeaturedBadge[];
}
